File: app/src/main/java/com/gmail/fattazzo/meteo/preferences/iconthemepicker/IconThemeListPreferenceAdapter.java
package com.gmail.fattazzo.meteo.preferences.iconthemepicker;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.TextView;

import com.gmail.fattazzo.meteo.R;
import com.gmail.fattazzo.meteo.utils.Closure;
import com.gmail.fattazzo.meteo.utils.icons.WeatherIconsFactory;
import com.gmail.fattazzo.meteo.utils.icons.retirever.IconsRetriever;

import java.util.List;

/**
 * @author fattazzo
 * <p/>
 * date: 23/02/16
 */
public class IconThemeListPreferenceAdapter extends ArrayAdapter<IconThemeItem> {

    private static final String TAG = "IconThemeAdapter";

    private Context context;
    private List<IconThemeItem> icons;
    private int resource;
    private Closure<Void> onThemeSelectedClosure;

    public IconThemeListPreferenceAdapter(Context context, int resource,
                                          List<IconThemeItem> objects) {
        super(context, resource, objects);
        this.context = context;
        this.resource = resource;
        this.icons = objects;
    }

    /**
     * Setter of onThemeSelectedClosure
     *
     * @param onThemeSelectedClosure the onThemeSelectedClosure to set
     */
    public void setOnThemeSelectedClosure(Closure<Void> onThemeSelectedClosure) {
        this.onThemeSelectedClosure = onThemeSelectedClosure;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {


        IconThemeViewHolder holder;
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(resource, parent, false);

            holder = new IconThemeViewHolder();
            holder.setThemeName((TextView) convertView
                    .findViewById(R.id.icon_theme_name));
            holder.setImage1((ImageView) convertView
                    .findViewById(R.id.icon_theme_image1));
            holder.setImage2((ImageView) convertView
                    .findViewById(R.id.icon_theme_image2));
            holder.setImage3((ImageView) convertView
                    .findViewById(R.id.icon_theme_image3));
            holder.setRadioButton((RadioButton) convertView
                    .findViewById(R.id.icon_theme_radio));

            convertView.setTag(holder);
        } else {
            holder = (IconThemeViewHolder) convertView.getTag();
        }

        holder.getThemeName().setText(icons.get(position).getName());

        IconsRetriever iconsRetriever = WeatherIconsFactory.INSTANCE.getIconsRetriever(icons.get(position).getPath());

        holder.getImage1().setImageResource(iconsRetriever.getIconDemo1());
        holder.getImage2().setImageResource(iconsRetriever.getIconDemo2());
        holder.getImage3().setImageResource(iconsRetriever.getIconDemo3());

        holder.getRadioButton().setChecked(icons.get(position).isChecked());

        convertView.setOnClickListener(v -> {
            for (int i = 0; i < icons.size(); i++) {
                if (i == position)
                    icons.get(i).setChecked(true);
                else
                    icons.get(i).setChecked(false);
            }
            if (onThemeSelectedClosure != null) {
                onThemeSelectedClosure.execute(null);
            }
        });

        return convertView;
    }
}

File: app/src/main/java/com/gmail/fattazzo/meteo/preferences/view/iconthemepicker/IconThemeListPreferenceAdapter.kt
/*
 * Project: meteo
 * File: IconThemeListPreferenceAdapter.kt
 *
 * Created by fattazzo
 * Copyright © 2019 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.meteo.preferences.view.iconthemepicker

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.RadioButton
import android.widget.TextView
import com.gmail.fattazzo.meteo.R
import com.gmail.fattazzo.meteo.utils.icons.WeatherIconsFactory

/**
 * @author fattazzo
 *
 *
 * date: 23/02/16
 */
class IconThemeListPreferenceAdapter(
    context: Context,
    private val resource: Int,
    private val icons: List<IconThemeItem>
) : ArrayAdapter<IconThemeItem>(context, resource, icons) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val convertViewAdapter: View

        val holder: IconThemeViewHolder
        if (convertView == null) {
            val inflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            convertViewAdapter = inflater.inflate(resource, parent, false)

            holder = IconThemeViewHolder()
            holder.themeName = convertViewAdapter
                .findViewById<View>(R.id.icon_theme_name) as TextView
            holder.image1 = convertViewAdapter
                .findViewById<View>(R.id.icon_theme_image1) as ImageView
            holder.image2 = convertViewAdapter
                .findViewById<View>(R.id.icon_theme_image2) as ImageView
            holder.image3 = convertViewAdapter
                .findViewById<View>(R.id.icon_theme_image3) as ImageView
            holder.radioButton = convertViewAdapter
                .findViewById<View>(R.id.icon_theme_radio) as RadioButton

            convertViewAdapter.tag = holder
        } else {
            convertViewAdapter = convertView
            holder = convertViewAdapter.tag as IconThemeViewHolder
        }

        holder.themeName!!.text = icons[position].name

        val iconsRetriever = WeatherIconsFactory.getIconsRetriever(icons[position].path)

        holder.image1!!.setImageResource(iconsRetriever.getIconDemo1())
        holder.image2!!.setImageResource(iconsRetriever.getIconDemo2())
        holder.image3!!.setImageResource(iconsRetriever.getIconDemo3())

        holder.radioButton!!.isChecked = icons[position].isChecked

        convertViewAdapter.setOnClickListener {
            for (i in icons.indices) {
                icons[i].isChecked = i == position
            }
            notifyDataSetChanged()
        }

        return convertViewAdapter
    }
}